Sally7/Internal/WireFormatting.cs

```diff
@@ -0,0 +1,69 @@
+﻿using System;
+using System.Buffers.Binary;
+using System.Runtime.CompilerServices;
+using Sally7.Protocol.S7.Messages;
+using Sally7.RequestExecutor;
+
+namespace Sally7.Internal;
+
+internal static class WireFormatting
+{
+    private const uint Tpkt = 0x03_00_00_00;
+
+    /// <summary>
+    /// Data struct.
+    /// </summary>
+    /// <remarks>
+    /// This struct is only 3 bytes long, final byte is overwritten.
+    /// </remarks>
+    private const int Data = 2 << 24 // Length
+        | 0b1111_0000 << 16 // Identifier
+        | 0b1000_0000 << 8; // PDU number and EOT
+
+    /// <summary>
+    /// The PDU reference.
+    /// </summary>
+    /// <remarks>
+    /// This value is here for legacy reasons. The <see cref="ConcurrentRequestExecutor"/> modifies the lower
+    /// byte to identify responses, the upper byte isn't actually used at the moment.
+    /// </remarks>
+    private const ushort PduRef = 1 << 8;
+
+    private static ulong JobRequestHeader1 = 0x32L << 56 | (long)MessageType.JobRequest << 48 | PduRef << 16;
+
+    public static uint WriteTpkt(ref byte destination, int length)
+    {
+        return WriteUInt32(ref destination, (uint) (Tpkt | length));
+    }
+
+    public static uint WriteData(ref byte destination)
+    {
+        // For performance we write an int instead of separate values, the final byte of the int will be overwritten.
+        WriteUInt32(ref destination, Data);
+
+        return 3;
+    }
+
+    public static uint WriteJobRequestHeader(ref byte destination, int paramLength, int dataLength)
+    {
+        var header = JobRequestHeader1 | (ushort) paramLength;
+
+        Unsafe.WriteUnaligned(ref destination,
+            BitConverter.IsLittleEndian ? BinaryPrimitives.ReverseEndianness(header) : header);
+        WriteUInt16(ref destination.GetOffset(8), (ushort)dataLength);
+
+        return 10;
+    }
+
+    public static uint WriteUInt16(ref byte destination, ushort value)
+    {
+        NetworkOrderSerializer.WriteUInt16(ref destination, value);
+        return sizeof(short);
```

Comment:
Is there something to access the compile time type of a variable in C# - which has no overhead?

Then you could just `return sizeof(decltype(value));` - making the code more generic with less chance of errors.

(I don't know why you currently don't use `sizeof(ushort)` - instead of adding the additional mental burden to know that sizeof for the same signed and unsigned type is the same)

Reply:
This started out as `short value`, thus return `sizeof(short)`. Forgot to clean these up as well. Unfortunately it's not possible to do `sizeof(value)`, but I bet there's quite some downsides to having that support in C#.